service/history/api/respondworkflowtaskcompleted/api.go

```diff
@@ -429,26 +429,26 @@ func (handler *WorkflowTaskCompletedHandler) Invoke(
 			wtFailedShouldCreateNewTask ||
 			bufferedEventShouldCreateNewTask ||
 			activityNotStartedCancelled {
+
 			newWorkflowTaskType = enumsspb.WORKFLOW_TASK_TYPE_NORMAL
 
+		} else if updateRegistry.HasOutgoingMessages(true) {
 			// There shouldn't be any sent updates in the registry because
 			// all sent but not processed updates were rejected by server.
 			// Therefore, it doesn't matter if to includeAlreadySent or not.
-		} else if updateRegistry.HasOutgoingMessages(true) {
-			if completedEvent == nil || ms.GetNextEventID() == completedEvent.GetEventId()+1 {
-				newWorkflowTaskType = enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE
-			} else {
-				newWorkflowTaskType = enumsspb.WORKFLOW_TASK_TYPE_NORMAL
-			}
+
+			newWorkflowTaskType = enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE
 		}
 	}
 
 	bypassTaskGeneration := request.GetReturnNewWorkflowTask() && wtFailedCause == nil
-	// TODO (alex-update): Need to support case when ReturnNewWorkflowTask=false and WT.Type=Speculative.
-	// In this case WT needs to be added directly to matching.
-	// Current implementation will create normal WT.
+	// TODO (alex-update): All current SDKs always set ReturnNewWorkflowTask to true
+	//  which means that server always bypass task generation if WFT didn't fail.
+	//  ReturnNewWorkflowTask flag needs to be removed.
+
 	if newWorkflowTaskType == enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE && !bypassTaskGeneration {
```

Comment:
I'm a big fan of creating a variable for a compound bool expression that summarizes it. For example here, could we use `needToConvertSpeculativeTaskToNormal` or something similar?

Reply:
I generally agree. In this case though, I assigned `ReturnNewWorkflowTask` field removal task to myself for the next sprint. It all will be gone very soon.